components/browser/session/src/main/java/mozilla/components/browser/session/SessionObserver.kt

```diff
@@ -0,0 +1,58 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.session
+
+/**
+ * This class is a combination of [Session.Observer] and
+ * [SessionManager.Observer]. It provides functionality to observe changes to a
+ * specified or selected session, and automatically takes care of switching
+ * over the observer in case a different session is selected (see [observe] and
+ * [observeSelected]).
+ *
+ * @property activeSession the currently observed session
+ * @property sessionManager the application's session manager
+ */
+abstract class SessionObserver(
+    private val sessionManager: SessionManager
+) : SessionManager.Observer, Session.Observer {
+
+    protected open var activeSession: Session? = null
+
+    /**
+     * Starts observing changes to the specified session.
+     *
+     * @param session the session to observe.
+     */
+    fun observe(session: Session) {
+        activeSession = session
+        session.register(this)
+    }
+
+    /**
+     * Starts observing changes to the selected session (see
+     * [SessionManager.selectedSession]). If a different session is selected
+     * the observer will automatically be switched over and only notified of
+     * changes to the newly selected session.
+     */
+    fun observeSelected() {
```

Comment:
API nit: I wonder if we can make it even more obvious that this is not only observing the *currently* selected session but updates? `observeSelection()`, `alwaysObserveSelected()`, ... not sure.